### DISCUSSION TITLE
Next Application Api Routes return 404

### DISCUSSION THREAD

**Discussion by Lightgaming** [vercel/vercel#5686] [Help]
Hello!

I have a Next.js application with dynamic API routes (check out next-auth.js.org is very nice btw.) 
Everything works fine when I run my project locally, but when its deployed on Vercel the routes return a 404. 

I also tested this with a non-dynamic route /pages/api/url.ts which worked fine locally, but also returned 404 on Vercel. 

I did look at the #5256 Post which didn't help either. (I also had an api folder next to the pages folder, but now it's called backend)

This is my folder structure:
![image](https://user-images.githubusercontent.com/18364007/104380349-29e8a800-552b-11eb-93de-248423cff226.png)

This is my next.config.js:

        const withPlugins = require('next-compose-plugins');
        const withFonts = require('next-fonts');
        const withTM = require('next-transpile-modules');
        const nextConfig = {
            poweredByHeader: false,
            distDir: 'build',
            transpileOnly: true,
            skipTypescriptChecking: true,
        
            typescript: {
                ignoreDevErrors: true,
                ignoreBuildErrors: true,
            },
        };
        module.exports = withPlugins(
            [[withTM], [withFonts]],
            nextConfig,
        );
 
I have a standard npm run build command: next build

Here is my latest build site: https://self-service-3t8n0c3as.vercel.app/
There should be these dynamic api routes available: https://next-auth.js.org/getting-started/rest-api
And a static route at https://self-service-3t8n0c3as.vercel.app/api/url

All of them return 404 :cry: 

Please help me

**Comment by joshkay** [comment]
Did you ever figure out what caused this?  I am running into the same issue.

**Comment by ajjack50n** [comment]
I have this problem too. All of my API routes return 404 but work perfectly on my dev environment.

**Reply by leerob** [reply]
Did you add the necessary environment variables inside Vercel for your production environment?

**Comment by Lightgaming** [comment]
@joshkay @ajjack50n  I'm sorry, but I can't really remember any more (I also don't have access to my code) but I think it was a problem with a limitation from Vercel on dynamic routes. 

I think this problem is correlated to the Serverless Function- and Dynamic Routes limitations.
https://vercel.com/docs/limits/overview#general-limits

I hope this kinda helps 😅
